recipes/3p_integrations/llamaindex/dlai_agentic_rag/Building_Agentic_RAG_with_Llamaindex_L4_Building_a_Multi-Document_Agent.ipynb

```diff
@@ -0,0 +1,4418 @@
+   },
+   "outputs": [],
+   "source": [
+    "from pathlib import Path\n",
+    "\n",
+    "paper_to_tools_dict = {}\n",
+    "for paper in papers:\n",
+    "    print(f\"Getting tools for paper: {paper}\")\n",
+    "    vector_tool, summary_tool = get_doc_tools(paper, Path(paper).stem)\n",
+    "    paper_to_tools_dict[paper] = [vector_tool, summary_tool]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "E2wXRc15Mp5Y"
+   },
+   "outputs": [],
+   "source": [
+    "all_tools = [t for paper in papers for t in paper_to_tools_dict[paper]]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "PKd8_J1vMp8a"
+   },
+   "outputs": [],
+   "source": [
+    "# define an \"object\" index and retriever over these tools\n",
+    "from llama_index.core import VectorStoreIndex\n",
+    "from llama_index.core.objects import ObjectIndex\n",
+    "\n",
+    "obj_index = ObjectIndex.from_objects(\n",
+    "    all_tools,\n",
+    "    index_cls=VectorStoreIndex,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "jys0qlogM069"
+   },
+   "outputs": [],
+   "source": [
+    "obj_retriever = obj_index.as_retriever(similarity_top_k=3)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "q6iWXlbsM3LX"
+   },
+   "outputs": [],
+   "source": [
+    "tools = obj_retriever.retrieve(\n",
+    "    \"Tell me about the eval dataset used in MetaGPT and SWE-Bench\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "oWIPUar2M4yp",
+    "outputId": "ad06205d-1996-4d6c-8a95-3c13824c75bd"
+   },
+   "outputs": [],
+   "source": [
+    "tools[2].metadata"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "qsD37iTpM6gb"
+   },
+   "outputs": [],
+   "source": [
+    "# The LlamaIndex's FunctionCallingAgentWorker API doesn't work correctly with Fireworks Llama, so we use ReActAgent here.\n",
+    "from llama_index.core.agent import ReActAgent\n",
+    "\n",
+    "agent = ReActAgent.from_tools(\n",
+    "    tool_retriever=obj_retriever,\n",
+    "    llm=llm,\n",
+    "    system_prompt=\"\"\" \\\n",
+    "You are an agent designed to answer queries over a set of given papers.\n",
+    "Please always use the tools provided to answer a question. Do not rely on prior knowledge.\\\n",
+    "\"\"\",    \n",
+    "    verbose=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 1000
+    },
+    "id": "M6gM7iq6M8De",
+    "outputId": "392e2376-1aff-46bb-ab23-1da137fa27ed"
+   },
+   "outputs": [],
+   "source": [
+    "response = agent.query(\n",
```

Comment:
Here it looks like it is able to parse information about the evaluation dataset, and identifies that it needs to call `vector_tool_swebench`, calls it and identifies what the dataset contains, but then fails to parse the output:

`Observation: Error: Could not parse output. Please follow the thought-action-input format. Try again.`
.....


`ValueError: Reached max iterations.`

Reply:
This could be caused by the occasional unreliable run of agent. If you run it again do you still see the error?

Reply:
Looks like it worked on a re-run.